astropy/modeling/core.py

```diff
@@ -866,13 +870,60 @@ def __setattr__(self, attr, value):
             else:
                 super().__setattr__(attr, value)
 
-    def __call__(self, *inputs, **kwargs):
+    def __call__(self, *args, **kwargs):
         """
         Evaluate this model using the given input(s) and the parameter values
         that were specified when the model was instantiated.
         """
+        def _keyword2positional(kwargs):
+            # these are the keys that are always present as keyword arguments
+            keys = ['model_set_axis', 'with_bounding_box', 'fill_value',
+                    'equivalencies', 'inputs_map']
+
+            new_inputs = {}
+            # kwargs contain the names of the new inputs + ``keys``
+            allkeys = list(kwargs.keys())
+            # Remove the names of the new inputs from kwargs and save them
+            # to a dict ``new_inputs``.
+            for key in allkeys:
+                if key not in keys:
+                    new_inputs[key] = kwargs[key]
+                    del kwargs[key]
+
+            return new_inputs
+
+        if not args:
+            # Inputs were passed as keyword (not positional) arguments.
+            # Because the signature of the ``__call__`` is defined at
+            # the class level, the name of the inputs cannot be changed at
+            # the instance level and the old names are always present in the
+            # signature of the method. In order to use the new names of the
+            # inputs, the old names are taken out of ``kwargs``, the input
+            # values are sorted in the order of self.inputs and passed as
+            # positional arguments to ``__call__``.
+
+            new_inputs = _keyword2positional(kwargs)
+            # Create positional arguments from the keyword arguments in ``new_inputs``.
+            new_args = []
+            for k in self.inputs:
+                new_args.append(new_inputs[k])
+        elif len(args) != self.n_inputs:
+            # Some inputs are passed as positional, others as keyword arguments.
+            args = list(args)
+
+            # Create positional arguments from the keyword arguments in ``new_inputs``.
+            new_inputs = _keyword2positional(kwargs)
+            new_args = []
+            for k in self.inputs:
+                if k in new_inputs:
+                    new_args.append(new_inputs[k])
+                else:
+                    new_args.append(args[0])
+                    del args[0]
```

Comment:
Suppose there are 3 inputs x, y, z, and this call is made f(1, 2, x=3, z=-1), it looks like to me what will happen is that the new_args list will be [2,  1,  -1], which is not what I would expect (as it goes through new_inputs, it does find x, so that value goes on the list, doesn't find y, so the first value of args goes on the list, then the value for z goes n the list.) So two possible issues: should it be an error if a values is specified twice, or do we take the keyword value? Did I say two? OK, one I guess. The other part seems like an error in that it can get the two out of synch when a positional argument is also provided as a keyword argument. Perhaps  split up the new_inputs into two keyword lists: one that corresponds to the length of args and the other which is the rest. If one of the new keywords is in the first, its an error, and if not, just append to args. But I may be confused.